src/test/storm/mesos/MesosNimbusTest.java

```diff
@@ -0,0 +1,509 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package storm.mesos;
+
+import backtype.storm.generated.StormTopology;
+import backtype.storm.scheduler.Topologies;
+import backtype.storm.scheduler.TopologyDetails;
+import backtype.storm.scheduler.WorkerSlot;
+import org.apache.mesos.MesosSchedulerDriver;
+import org.apache.mesos.Protos.*;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import java.net.URI;
+import java.util.*;
+
+import static org.junit.Assert.assertEquals;
+
+public class MesosNimbusTest {
+
+  private Offer buildOffer(double cpus, double mem) {
+    return Offer.newBuilder()
+        .setId(OfferID.newBuilder().setValue("derp").build())
+        .setFrameworkId(FrameworkID.newBuilder().setValue("derp").build())
+        .setSlaveId(SlaveID.newBuilder().setValue("derp").build())
+        .setHostname("derp")
+        .addAllResources(
+            Arrays.asList(
+                buildScalarResource("cpus", cpus),
+                buildScalarResource("mem", mem)
+            )
+        )
+        .build();
+  }
+
+  private Offer buildOfferWithPorts(double cpus, double mem, int portBegin, int portEnd) {
+    return Offer.newBuilder()
+        .setId(OfferID.newBuilder().setValue("derp").build())
+        .setFrameworkId(FrameworkID.newBuilder().setValue("derp").build())
+        .setSlaveId(SlaveID.newBuilder().setValue("derp").build())
+        .setHostname("derp")
+        .addAllResources(
+            Arrays.asList(
+                buildScalarResource("cpus", cpus),
+                buildScalarResource("mem", mem),
+                buildRangeResource("ports", portBegin, portEnd)
+            )
+        )
+        .build();
+  }
+
+  private Offer buildOfferWithReservation(double cpus, double mem, double reservedCpu, double reservedMem) {
```

Comment:
Please add a test to offer resources with a dynamic reservation. With principals that you're allowed to use and that you're not allowed to use.